src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -530,61 +490,59 @@ internal void VerifyMutable()
             }
         }
 
-        private volatile bool _isConfigured;
-        private readonly object _configureLock = new object();
+        internal bool IsConfigured => _configureState == 2;
+        private volatile int _configureState; // 0: not configured, 1: configuring, 2: configured
         private ExceptionDispatchInfo? _cachedConfigureError;
 
-        internal bool IsConfigured => _isConfigured;
-
         internal void EnsureConfigured()
         {
-            Debug.Assert(!Monitor.IsEntered(_configureLock), "recursive locking detected.");
-
-            if (!_isConfigured)
-                ConfigureLocked();
+            if (!IsConfigured)
+                ConfigureSynchronized();
 
-            void ConfigureLocked()
+            void ConfigureSynchronized()
             {
+                Options.MakeReadOnly();
+                MakeReadOnly();
+
                 _cachedConfigureError?.Throw();
 
-                lock (_configureLock)
+                lock (Options.CacheContext)
```

Comment:
Eager traversal means that the previous locking strategy could result in deadlocks, so we need to replace it with a lock on the entire metadata cache.

Cf. #81199

Reply:
is this the cache context shared between the equivalent options? (I'm talking about that piece of logic which causes that options instance doesn't have to be the same for different type infos when the options are equivalent)

Reply:
That's right. This is why need to lock on the shared instance and not on the options since that could result in races.